pkg/runcfanotify/runcfanotify.go

```diff
@@ -0,0 +1,352 @@
+// or terminated, and call the callback on such event.
+//
+// Limitations:
+// - runc must be installed in one of the paths listed by runcPaths
+// - runc must be executed by containerd-shim
+// - Linux >= 5.3 (for pidfd_open)
+func NewRuncNotifier(callback RuncNotifyFunc) (*RuncNotifier, error) {
+	n := &RuncNotifier{
+		stop:     make(chan struct{}),
+		callback: callback,
+	}
+
+	fanotifyFlags := uint(unix.FAN_CLOEXEC | unix.FAN_CLASS_CONTENT | unix.FAN_UNLIMITED_QUEUE | unix.FAN_UNLIMITED_MARKS)
+	openFlags := os.O_RDONLY | unix.O_LARGEFILE | unix.O_CLOEXEC
+
+	notify, err := fanotify.Initialize(fanotifyFlags, openFlags)
+	if err != nil {
+		return nil, err
+	}
+	n.notify = notify
+
+	for _, file := range runcPaths {
+		err = notify.Mark(unix.FAN_MARK_ADD, unix.FAN_OPEN_EXEC_PERM, unix.AT_FDCWD, file)
+		if err == nil {
+			log.Debugf("Checking %q: done", file)
+		} else {
+			log.Debugf("Checking %q: %s", file, err)
+		}
+	}
+
+	go n.run()
+
+	return n, nil
+}
+
+func monitorRuncInstance(bundleDir string, pidFile string, callback RuncNotifyFunc) error {
+	fanotifyFlags := uint(unix.FAN_CLOEXEC | unix.FAN_CLASS_CONTENT | unix.FAN_UNLIMITED_QUEUE | unix.FAN_UNLIMITED_MARKS)
+	openFlags := os.O_RDONLY | unix.O_LARGEFILE | unix.O_CLOEXEC
+
+	notify, err := fanotify.Initialize(fanotifyFlags, openFlags)
+	if err != nil {
+		return err
+	}
+
+	pidFileDir := filepath.Dir(pidFile)
+	log.Infof("runcfanotify: mark %s\n", pidFileDir)
+	err = notify.Mark(unix.FAN_MARK_ADD, unix.FAN_ACCESS_PERM|unix.FAN_EVENT_ON_CHILD, unix.AT_FDCWD, pidFileDir)
+	if err != nil {
+		log.Infof("Cannot mark %s: %s", bundleDir, err)
+	}
+	configJsonPath := filepath.Join(bundleDir, "config.json")
+	log.Infof("runcfanotify: mark ignore %s", configJsonPath)
+	err = notify.Mark(unix.FAN_MARK_ADD|unix.FAN_MARK_IGNORED_MASK, unix.FAN_ACCESS_PERM, unix.AT_FDCWD, configJsonPath)
+	if err != nil {
+		log.Infof("Cannot ignore %s/config.json: %s\n", bundleDir, err)
+	}
+
+	containerID := filepath.Base(filepath.Clean(bundleDir))
+
+	// f returns containerID, containerPID, error
+	f := func(notify *fanotify.NotifyFD, callback RuncNotifyFunc) (bool, error) {
+		data, err := notify.GetEvent()
+		if err != nil {
+			return false, fmt.Errorf("%w", err)
+		}
+
+		if data == nil {
+			return false, nil
+		}
+
+		defer data.Close()
+		if !data.MatchMask(unix.FAN_ACCESS_PERM) {
+			return false, fmt.Errorf("fanotify: unknown event")
+		}
+		defer notify.ResponseAllow(data)
+
+		pid := data.GetPID()
+
+		// Cannot use GetEvent(skipPIDs) because we need to call ResponseAllow
+		if pid == os.Getpid() {
+			return false, nil
+		}
+
+		path, err := data.GetPath()
+		if err != nil {
+			return false, err
+		}
+		if path != pidFile {
+			return false, nil
+		}
+
+		commB, err := ioutil.ReadFile(fmt.Sprintf("/proc/%d/comm", pid))
+		if err != nil {
+			return false, err
+		}
+		comm := strings.TrimSuffix(string(commB), "\n")
+
+		if comm != "containerd-shim" {
+			return false, nil
+		}
```

Comment:
I'm curious why it's a requirement?. Is it because the source mounts should be set in a specific way to create the runc container to kubernetes container relationship?

Reply:
It was not about the mounts but because of the sequence of creating a container:
1. containerd-shim executes "runc create" to create the container
2. runc opens the pid file
3. runc writes the pid in the pid file
4. The command "runc create" terminates
5. containerd-shim opens the pid file
6. containerd-shim reads the pid from the pid file
7. containerd-shim executes "runc start" to start the container

Here I wanted to pause on step 5 because at this stage the pid file is already created with the right content. Without the test `comm != "containerd-shim"`, it might be that the event triggered at step 2 and it would not have the right content yet.

But I can remove this test and replace it by a check that the file is not empty. If it is empty, it means the event is triggered by something else and I should ignore the event.

Reply:
I'm asking because I'm wondering if we can support other runtimes than containerd-shim?

Reply:
I guess it's possible but it is not tested...